server/scsynth/SC_PortAudio.cpp

```diff
@@ -262,39 +270,181 @@ int SC_PortAudioDriver::PortAudioCallback(const void* input, void* output, unsig
     return paContinue;
 }
 
-void SC_PortAudioDriver::GetPaDeviceFromName(const char* device, int* mInOut) {
-    const PaDeviceInfo* pdi;
-    const PaHostApiInfo* apiInfo;
-    char devString[256];
-    PaDeviceIndex numDevices = Pa_GetDeviceCount();
-    mInOut[0] = paNoDevice;
-    mInOut[1] = paNoDevice;
+std::string SC_PortAudioDriver::GetPaDeviceName(int index) const {
+    auto* pdi = Pa_GetDeviceInfo(index);
+    std::string name;
+#ifndef __APPLE__
+    name += Pa_GetHostApiInfo(pdi->hostApi)->name;
+    name += " : ";
+#endif
+    name += pdi->name;
+    return name;
+}
+
+void SC_PortAudioDriver::GetPaDeviceFromName(const char* device, int* mInOut, IOType ioType) const {
+    if (device[0] == '\0')
+        return;
 
-    // This tries to find one or two devices that match the given name (substring)
-    // might cause problems for some names...
+    PaDeviceIndex numDevices = Pa_GetDeviceCount();
+    *mInOut = paNoDevice;
     for (int i = 0; i < numDevices; i++) {
-        pdi = Pa_GetDeviceInfo(i);
-        apiInfo = Pa_GetHostApiInfo(pdi->hostApi);
-        strcpy(devString, apiInfo->name);
-        strcat(devString, " : ");
-        strcat(devString, pdi->name);
-        if (strstr(devString, device)) {
-            if (pdi->maxInputChannels > 0)
-                mInOut[0] = i;
-            if (pdi->maxOutputChannels > 0)
-                mInOut[1] = i;
+        auto* pdi = Pa_GetDeviceInfo(i);
+        std::string devString = GetPaDeviceName(i);
+        if (strstr(devString.c_str(), device)) {
+            if (ioType == IOType::Input && pdi->maxInputChannels > 0) {
+                *mInOut = i;
+                break;
+            } else if (ioType == IOType::Output && pdi->maxOutputChannels > 0) {
+                *mInOut = i;
+                break;
+            }
```

Comment:
OK, sorry for iterating but it took some time to see what this function is actually trying to do. Here is an idiomatic C++ way to write it:

```cpp

int SC_PortAudioDriver::GetPaDeviceFromName(const char* device, IOType ioType) const {
    if (device == nullptr || device[0] == '\0')
        return paNoDevice;

    PaDeviceIndex numDevices = Pa_GetDeviceCount();
    for (int i = 0; i < numDevices; i++) {
        auto* pdi = Pa_GetDeviceInfo(i);
        std::string devString = GetPaDeviceName(i);
        if (strstr(devString.c_str(), device)) {
            if (ioType == IOType::Input && pdi->maxInputChannels > 0) {
                return i;
            } else if (ioType == IOType::Output && pdi->maxOutputChannels > 0) {
                return i;
            }
        }
    }

    return paNoDevice;
}
```

note:
- return value is actually returned from function rather than passed in via pointer
- null check moves inside this function
- clearer behavior when no device is found

this means that at the call site below you can directly write:

```cpp
    mDeviceInOut[0] = GetPaDeviceFromName(mWorld->hw->mInDeviceName, &mDeviceInOut[0], IOType::Input);
    mDeviceInOut[1] = GetPaDeviceFromName(mWorld->hw->mOutDeviceName, &mDeviceInOut[1], IOType::Output);
```

Reply:
This makes sense. As you observed, I was trying to follow existing paradigms, but this looks nicer for sure.